src/main/java/org/jenkinsci/plugins/workflow/actions/WarningAction.java

```diff
@@ -0,0 +1,37 @@
+package org.jenkinsci.plugins.workflow.actions;
+
+import hudson.model.Result;
+import javax.annotation.Nonnull;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+/**
+ * This action can be attached to a {@link FlowNode} to signify that some non-fatal
+ * warning occurred during execution. Visualizations should treat FlowNodes with this
+ * action as if the FlowNode's result was {@link Result#UNSTABLE}.
+ */
+public final class WarningAction implements PersistentAction {
+    private final @Nonnull String message;
+
+    public WarningAction(@Nonnull String message) {
```

Comment:
Maybe it would make more sense for this to be `@Nullable` so that something like `unstable()` with no message would work without having to create some kind of default message.

Reply:
I do not see a use case for a null message, so would suggest keeping it `@Nonnull` unless and until one is found.

Reply:
On second thought, perhaps it would be appropriate to have the constructor take no arguments, then use the idiom

```java
public WarningAction withMessage(@Nonnull String message) {
    this.message = message;
    return this;
}
```

as I think it is likely we will at some point want an associated URL and perhaps even a badge icon.

Reply:
Any particular reason to prefer `withMessage`, `withUrl`, etc. over adding new constructors that take additional arguments? With the former pattern it is easy to forget to call `withMessage`. We could use the builder pattern and validate that the message is non-null when constructing the `WarningAction` from  the builder, but that seems like overkill.

Reply:
The builder style makes it easier to introduce new features compatibly. If we decide that a message is mandatory, then keep it in the constructor.